### ISSUE TITLE
"ReferenceError: __extends is not defined" in tests

### ISSUE THREAD

**Issue by jplhomer** [jsonapi-suite/jsorm#31]
When using `mocha`, `mocha-webpack` and `jsdom` to run tests, we see an error when testing a JS file that imports a model from `jsorm`. We're using Vanilla JS + Vue loader.

```
ReferenceError: __extends is not defined
    at /playlist-admin/node_modules/jsorm/_bundles/webpack:/src/request.ts:7:28
    at Object.<anonymous> (/playlist-admin/node_modules/jsorm/_bundles/webpack:/src/request.ts:18:2)
    at __webpack_require__ (/playlist-admin/node_modules/jsorm/_bundles/webpack:/webpack/bootstrap 96ad672db14dcf69449e:19:1)
    at Object.__webpack_exports__.a.target.__attrDecorators (/playlist-admin/node_modules/jsorm/_bundles/jsorm.js:2615:67)
    at __webpack_require__ (/playlist-admin/node_modules/jsorm/_bundles/webpack:/webpack/bootstrap 96ad672db14dcf69449e:19:1)
    at Object.<anonymous> (/playlist-admin/node_modules/jsorm/_bundles/jsorm.js:279:65)
    at __webpack_require__ (/playlist-admin/node_modules/jsorm/_bundles/webpack:/webpack/bootstrap 96ad672db14dcf69449e:19:1)
    at Object.<anonymous> (/playlist-admin/node_modules/jsorm/_bundles/jsorm.js:2470:65)
    at __webpack_require__ (/playlist-admin/node_modules/jsorm/_bundles/webpack:/webpack/bootstrap 96ad672db14dcf69449e:19:1)
    at Object.<anonymous> (/playlist-admin/node_modules/jsorm/_bundles/jsorm.js:2280:64)
    at __webpack_require__ (/playlist-admin/node_modules/jsorm/_bundles/webpack:/webpack/bootstrap 96ad672db14dcf69449e:19:1)
    at Object.<anonymous> (/playlist-admin/node_modules/jsorm/_bundles/jsorm.js:7745:18)
```

(the stack trace goes on for some time).

I'm seeing [similar references](https://github.com/akveo/ng2-smart-table/issues/16) across the web, and it seems to be related to `noEmitHelpers` being enabled in `tsconfig`. I'm not familiar with Typescript - so I'm hoping y'all know if this can be changed to fix our issue, or if there's another workaround.

Thanks!

**Comment by richmolj** [comment]
Hey @jplhomer - first, we'll have a big typescript overhaul coming very shortly that will prevent issues like this https://github.com/jsonapi-suite/jsorm/pull/30

Second, I think this would fix your issue in the meantime:

```
// test/tshook.js
require("ts-node").register({
  compilerOptions: {
    module: "commonjs",
    noEmitHelpers: false
  }
});
```

And run tests with `mocha -r test/tshook.js`. Does that help?

**Comment by jplhomer** [comment]
@richmolj Thanks for the update! 

I tried adding the `ts-node` hook into my Mocha setup script without success. Still getting the complaint about `__extends`. I also tried adding it as a global method - no luck there either.

I was unable to repro this issue in a new project with just Webpack + Mocha + jsorm. I'm wondering if this is getting caught up somewhere in the `vue-loader` logic, which is out of this project's wheelhouse.

I'll consider some other options - thanks again for your help.

**Comment by richmolj** [comment]
/cc @wadetandy who might have some ideas as well.

Though I don't think this will get you there, I'll note `patchExtends`:

```ts
import { patchExtends } from 'jsorm'
patchExtends()
```

Will generate that `__extend` method. You shouldn't ever need to call it directly though.

For context, this is because Typescript and ES6 classes don't have an `inherited` hook. So, in the ES5 transpilation where TS usually generates that `__extends` method, we generate our own. That's why the `noEmitHelpers` - we're emitting our own custom helper.

That said, it's not ideal and @wadetandy has been working to eliminate this issue altogether with the use of es7 decorators.

**Comment by wadetandy** [comment]
@jplhomer just to clarify, is the project you are using jsorm from also a typescript project or is it just JavaScript?

Also, do you still have the issue if you downgrade to jsorm 0.4.0?

**Comment by jplhomer** [comment]
@wadetandy using just JavaScript + Vue - no typescript.

I tried downgrading to 0.4.0 but the same error persists, unfortunately.

**Comment by jplhomer** [comment]
OK, I think I've narrowed the cause down to a conflict with a dependency for Vue testing: [`jsdom-global`](https://github.com/rstacruz/jsdom-global). It's added in my Mocha setup script, and it injects DOM-like objects into the Node global object:

```js
var document = new jsdom.JSDOM(html, options)
var window = document.window
global.document = window.document
global.window = window
// etc
```

This tricks `jsorm` into thinking it is running in a browser context, when it's actually running in a Node context. Here's an example of the `jsorm` library compiled to JS:

```js
/* WEBPACK VAR INJECTION */(function(global) {// Override defined here w/noEmitHelpers
// https://github.com/Microsoft/TypeScript/issues/6425
// Allows 'inherited' hook
var globalObj;
if (typeof window === 'undefined') {
    globalObj = global;
}
else {
    globalObj = window;
}
```

I can hardcode a fix in the compiled JS where the global/window check is performed, and my issue is resolved!

```diff
/* WEBPACK VAR INJECTION */(function(global) {// Override defined here w/noEmitHelpers
// https://github.com/Microsoft/TypeScript/issues/6425
// Allows 'inherited' hook
var globalObj;
-if (typeof window === 'undefined') {
+if (typeof window === 'undefined' || window !== global) {
    globalObj = global;
}
else {
    globalObj = window;
}
```

I have a fix written in the source repo, but Typescript is throwing a compiler error:

```ts
let globalObj;
if (typeof window === 'undefined' || window != global) {
  globalObj = global;
} else {
  globalObj = window;
```

Results in:

```
TSError: ⨯ Unable to compile TypeScript
src/custom-extend.ts (6,38): Operator '!==' cannot be applied to types 'Window' and 'Global'. (2365)
```

Any thoughts on how we could compare the injected global scope with the `window` var?

I really appreciate your help!

**Comment by wadetandy** [comment]
Easiest way to solve this is probably to cast one or both of `window` or `global` as an `any` type:

``` typescript
let globalObj;
if (typeof window === 'undefined' || window !== (global as any)) {
  globalObj = global;
} else {
  globalObj = window;
}
```

If you can submit a patch for this, I can release a new version.  Otherwise I'll probably get around to it early next week.

**Comment by jplhomer** [comment]
Awesome - opened up a PR! https://github.com/jsonapi-suite/jsorm/pull/32

**Comment by wadetandy** [comment]
Fixed by #32. Have released https://github.com/jsonapi-suite/jsorm/releases/tag/v0.5.2 to include this fix.  Thanks @jplhomer!
